src/lib/xmlrpc.ml

```diff
@@ -14,131 +14,130 @@
  * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  *)
-
 open Printf
 open Rpc
 
 (* marshalling/unmarshalling code *)
-(* The XML-RPC is not very clear about what characters can be in a string value ... *)
-let encode =
-  let translate = function
-    | '>' -> Some "&gt;"
-    | '<' -> Some "&lt;"
-    | '&' -> Some "&amp;"
-    | '"' -> Some "&quot;"
-    | c when (c >= '\x20' && c <= '\xff') || c = '\x09' || c = '\x0a' || c = '\x0d' ->
-      None
-    | _ -> Some ""
-  in
-  Internals.encode translate
-
 
-let rec add_value ?(strict = false) f = function
-  | Null -> f "<value><nil/></value>"
+let[@inline always] encode_add buf s m ~i n =
+  Buffer.add_substring buf s m (i - m);
+  Buffer.add_string buf n;
+  i + 1
+
+
+(** Encodes a string using the given translation function that maps a character
+    to a string that is its encoded version, if that character needs encoding. *)
+let[@inline always] encode_buf buf s =
```

Comment:
A faster approach could be: go over the string and calculate the length of the encoded string by watching for characters that need encoding; then reserve the required string and fill it, doing the actual encoding. In the best case, nothing needs to be done and the existing string can be used as is. This would work best if encoding is rare. The current approach relies and the performance of `Buffer` and always creates a copy. Another approach could be to use `String.exists` to check if any character needs encoding -- then use the code here; otherwise skip this and use the string as is. That would be easy to implement here.

Reply:
Buffer is pretty performant though: https://discuss.ocaml.org/t/whats-the-fastest-way-to-concatenate-strings/2616/12?u=mseri

Reply:
There was some code for 'needs_encoding' here before, but it was slower (in the end you need to add the string to the buffer in both situations, and if no encoding needs to be done, then the loop will exit without making any modifications, and the final `add_substring` will copy).
I tried to write also an 'estimate_size' that'd go through all the RPCs to try to get an estimate upper bound for the string to allocate, although that was slower too (even if I didn't traverse the string, just looked at its length)

Although I think my benchmark doesn't cover the case where strings need to be encoded, I'll tweak the benchmark and then try these optimizations again.

Reply:
Scanning is always fast because it does not allocate. But good observation that in the positive case only one copy will happen.